empress/taxonomy_utils.py

```diff
@@ -0,0 +1,151 @@
+            # in the feature metadata, which will make distinguishing these
+            # column(s) from the columns we were going to add from feature
+            # metadata difficult.
+            raise TaxonomyError(
+                "The feature metadata contains a taxonomy column, but also "
+                "already contains column(s) starting with the text 'Level' "
+                "(case insensitive)."
+            )
+
+        max_sc_count = 0
+
+        def find_max_semicolon_count(fm_row):
+            # We need to be able to update max_sc_count as we go through each
+            # row, and Python gets angry if we don't explicitly declare it as
+            # "nonlocal" before referencing it here. See
+            # https://stackoverflow.com/a/46018922/10730311
+            nonlocal max_sc_count
+            row_sc_count = fm_row.loc[tax_col_name].count(";")
+            max_sc_count = max(max_sc_count, row_sc_count)
+
+        # Figure out maximum number of semicolons in any feature (this works
+        # on a per-row basis even though axis="columns", because each call
+        # to validate_semicolon_count() gets a Series of columns which is
+        # technically a row I guess)
+        feature_metadata.apply(find_max_semicolon_count, axis="columns")
+
+        if max_sc_count == 0:
+            # We allow this in the case of single-rank taxonomies (e.g. just
+            # kingdoms, for some reason). Can change this to an error if
+            # desired.
+            warnings.warn(
+                (
+                    "None of the taxonomy values in the feature metadata "
+                    "contain a semicolon (;). Please make sure your taxonomy "
+                    'is formatted so that "levels" are separated by '
+                    "semicolons."
+                ),
+                TaxonomyWarning
+            )
+
+        # OK, actually do splitting now (taking into account max_sc_count)
+        def split_taxonomy_col(fm_row):
+            levels = [r.strip() for r in fm_row.loc[tax_col_name].split(";")]
+            # If this row's taxonomy has less levels than the max number of
+            # levels, pad it out with empty strings.
+            if len(levels) < max_sc_count + 1:
+                num_missing_levels = (max_sc_count + 1) - len(levels)
+                levels += ["Unspecified"] * num_missing_levels
+            return levels
+
+        # Our use of result_type="expand" means that tax_levels will be a
+        # DataFrame with the same index as feature_metadata but with one column
+        # for each taxonomic level (in order -- Kingdom, Phylum, etc.)
+        tax_levels = feature_metadata.apply(
+            split_taxonomy_col, axis="columns", result_type="expand"
+        )
+        # Assign human-friendly column names: Level 1, Level 2, ...
+        tax_levels.columns = [
+            "Level {}".format(i) for i in range(1, len(tax_levels.columns) + 1)
+        ]
```

Comment:
I think the following snippet would use pandas' API a bit better, let me know what you think.

```suggestion
        max_tags =  feature_metadata[tax_col_name].str.count(';').max()
        
        # add 2 for the trailing name and the non-zero index start
        levels = pd.DataFrame(index=feature_metadata.index, dtype='str',
                              columns=['level %d' % i for i in range(1, max_tags + 2)])
        
        for i, names in enumerate(df.Taxon.str.split(';')):
            levels.iloc[i, :len(names)] = names
        
        levels.fillna('Unspecified', inplace=True)

```

Reply:
This is a more concise way of doing things, but I ran into some problems trying to implement this code and make it pass the tests -- it doesn't strip out internal whitespace around levels (so the output has things like ` p__Firmicutes`), and after fixing that for some reason `names` is being set as `nan` for some rows during the enumerate loop. (Also, I think my original code is somewhat easier to read, although I am likely biased since I wrote it...)

Since the original code works, for the sake of time I'm going to suggest we move on for now? I can revisit this in the future if you'd like.

Reply:
Let's touch base on slack, I think it shouldn't be that far off to strip trailing and leading spaces. Trying `str.strip()` might do it.